### DISCUSSION TITLE
How to add C++ host init function to python model

### DISCUSSION THREAD

**Discussion by jacksonsdean** [FLAMEGPU/FLAMEGPU2#961] [Q&A] [answered]
Is it possible to add a host function to a model description in python using SWIG?

for example:
```
init_fcn = r"""
    FLAMEGPU_INIT_FUNCTION(init_fcn) { 
        // init function
    }
"""

model = pyflamegpu.ModelDescription("example")
model.addInitFunctionCallback(init_fcn) # fails
```

Do I need to create a HostFunctionCallback instance, and if so how can I create that from the string?

**Comment by Robadob** [comment, marked as answer]
The Python API allows you to define Host functions using Python, the documentation for that (with examples) can be found here: https://docs.flamegpu.com/guide/host-functions/defining-host-functions.html

It's not possible to runtime compile host C++, the same way that we can for agent functions (device/CUDA C++). Therefore, to attach a C++ host function to a python model would require embedding that host function into `pyflamegpu` to make it available to all Python users. This solution isn't particularly scalable, so isn't something we've considered.

Is there a specific reason you feel the need for a C++ host function within a Python model?

**Reply by jacksonsdean** [reply]
That makes sense, thanks for the response. The reason we were hoping to be able to use a C++ host function in Python is that we have an existing C++ model that we are hoping to run some experiments on using Python, since we want to use libraries like PyTorch without having to compile LibTorch into the C++ model. 

Our init function in particular is quite long and complex and so we were hoping to be able to avoid porting it to Python, but it sounds like we may end up needing to go that route.

**Reply by Robadob** [reply]
As far as I recall, custom reduction/transform operators are the only HostAPI feature not supported by Python. If you aren't making use of those (which seems unlikely for an init function) you should be ok.
